2018/2018_File_Generation_Features.md

```diff
@@ -0,0 +1,67 @@
+NULIs (loan ID) of the rows that fail the edit. Each method creates a results dataframe that holds the rows that have failed the specific edit. The class then wraps the results into an edit report that can be parsed by other methods in the repository.  
+
+- `lar_constraints` in [`lar_constraints.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/rules_engine.py) contains methods for creating clean LAR rows that conform to pass syntax or validity edits. New edit logic methods must be made in terms of changing a TS or LAR row. 
+
+- `test_data` in [`test_file_generator.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/test_file_generator.py) contains methods that manipulate LAR or TS dataframes to fail a specific syntax or validity edit. New edit logic methods must be made in terms of changing a TS or LAR dataframe.
+
+Each edit logic method in `lar_constraints`, `test_data`, and `rules_engine` is prefaced with 's', 'v', or 'q' to indicate a syntax, validity or quality edit respectively.
+
+### Validating Quality Edit Files and Creating Edit Reports
+The `test_data` logic for creating quality edit test files may cause rows to fail not only quality edits, but also syntax or validity edits in the creation process. One method in the FileGenerator class, `validate_quality_edit_file(quality_filename)` takes a quality edit test file from the quality filepaths directory in `test_filepaths.yaml`, drops rows that have syntax or validity edits, and duplicates the remaining clean rows to the length of the original file. The file is then saved in a new directory.
+
+FilesGenerator also contains a method, `edit_report` that produces an edit report for a test file which is stored locally as a csv file. The edit report csv is stored in a directory and with a filename specified in `test_filepaths.yaml` 
+
+## The Large Test Files Script
+The Large Test Files script in [`large_test_files_script.py`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/large_test_files_script.py) creates a test file based on existing test file data to create a new file with a larger number of rows. The script would only need to be run, and the attributes of the new file to be created can be changed with [`large_file_script_config.yaml`](https://github.com/cfpb/hmda-test-files/blob/master/2018/python/configurations/large_file_script_config.yaml) in `python/configurations`. Files can be created with any number of rows, and can be made for any test institution so long as the Institution Name, LEI and Tax ID are provided in the configuration file. The configuration file stores filepaths for source data, as well as the filepaths for output. 
+
```

Comment:
Remove `in addition`
replace `using a yaml configuration` with `with this yaml configuration`.

Remove the second sentence.

Reply:
Sure - I implemented these changes in the discussion on the large test files script in the main readme.